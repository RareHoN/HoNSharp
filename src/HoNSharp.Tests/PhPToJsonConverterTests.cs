using ApprovalTests;
using ApprovalTests.Reporters;
using Newtonsoft.Json;
using HoNSharp.Converters;
using HoNSharp.Models;

namespace HoNSharp.Tests
{
    [UseReporter(typeof(DiffReporter))]
    public class PhPToJsonConverterTests
    {
        private readonly PhPToJsonConverter _converter;

        public PhPToJsonConverterTests()
        {
            _converter = new PhPToJsonConverter();
        }

        [Fact]
        public void ConvertToJson_GetAccountAllHeroStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_account_all_hero_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<AllHeroStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithAnotherError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_another_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithDateTimeOffsetError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_datetimeoffset_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_GetMatchStats_WithUnexpectedCharacterError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_match_stats_unexpected_character_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MatchStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewMidwars_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_midwars_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewNormal_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_normal_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_MatchHistoryOverviewEmpty_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("match_history_overview_empty_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<Dictionary<string, string>>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaign_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaign_WithLongConvertError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_long_convert_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaignError_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_error_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsCampaignMidwars_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_campaign_midwars_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<CampaignStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_ShowStatsMastery_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("show_stats_mastery_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<MasteryStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_SimpleStats_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("simple_stats_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<SimpleStatsResponse>(result);
        }

        [Fact]
        public void ConvertToJson_PlayerAwardSumm_ShouldCorrectlyConvert()
        {
            // Arrange
            var responseFileContent = ReadFile("get_player_award_summ_response.txt");

            // Act
            var result = _converter.ConvertToJson(responseFileContent);

            // Assert
            Approvals.Verify(FormatJson(result));

            _ = JsonConvert.DeserializeObject<AwardSummaryResponse>(result);
        }

        private string ReadFile(string responseFileName)
        {
            var responseFilePath = Path.Combine(Directory.GetCurrentDirectory(), "TestData", responseFileName);

            return File.ReadAllText(responseFilePath);
        }

        private string FormatJson(string json)
        {
            var deserialized = JsonConvert.DeserializeObject(json);
            return JsonConvert.SerializeObject(deserialized, Formatting.Indented);
        }
    }
}