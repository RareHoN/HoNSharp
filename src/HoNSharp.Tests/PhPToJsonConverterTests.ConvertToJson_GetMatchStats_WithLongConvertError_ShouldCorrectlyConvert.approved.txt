{
  "super_id": 36523,
  "nickname": "IDoiz69z",
  "standing": 3,
  "account_type": 4,
  "account_id": 36523,
  "level": 1,
  "level_exp": 0,
  "discos": 0,
  "possible_discos": 0,
  "games_played": 0,
  "num_bot_games_won": 0,
  "total_games_played": 0,
  "total_discos": 0,
  "acc_secs": 0,
  "acc_games_played": 0,
  "acc_discos": 0,
  "rnk_secs": 0,
  "rnk_games_played": 0,
  "rnk_discos": 0,
  "cs_secs": 0,
  "cs_games_played": 0,
  "cs_discos": 0,
  "mid_games_played": 0,
  "mid_discos": 0,
  "rift_games_played": 0,
  "rift_discos": 0,
  "last_activity": "9/12/2022",
  "create_date": "9/12/2022",
  "season": 22,
  "cam_games_played": 0,
  "cam_wins": 0,
  "cam_losses": 0,
  "cam_concedes": 0,
  "cam_concedevotes": 0,
  "cam_buybacks": 0,
  "cam_discos": 0,
  "cam_kicked": 0,
  "cam_amm_solo_rating": {
    "MinValue": {
      "MinValue": null,
      "Epsilon": {
        "MinValue": null,
        "Epsilon": null,
        "MaxValue": {
          "MinValue": null,
          "Epsilon": null,
          "MaxValue": null,
          "PositiveInfinity": {
            "MinValue": null,
            "Epsilon": null,
            "MaxValue": null,
            "PositiveInfinity": null,
            "NegativeInfinity": {
              "MinValue": null,
              "Epsilon": null,
              "MaxValue": null,
              "PositiveInfinity": null,
              "NegativeInfinity": null,
              "NaN": {
                "MinValue": null,
                "Epsilon": null,
                "MaxValue": null,
                "PositiveInfinity": null,
                "NegativeInfinity": null,
                "NaN": null
              }
            },
            "NaN": null
          },
          "NegativeInfinity": null,
          "NaN": null
        },
        "PositiveInfinity": null,
        "NegativeInfinity": null,
        "NaN": null
      },
      "MaxValue": null,
      "PositiveInfinity": null,
      "NegativeInfinity": null,
      "NaN": null
    },
    "Epsilon": null,
    "MaxValue": null,
    "PositiveInfinity": null,
    "NegativeInfinity": null,
    "NaN": null
  },
  "cam_amm_solo_count": 0,
  "cam_amm_solo_conf": 0,
  "cam_amm_solo_prov": 0,
  "cam_amm_solo_pset": 0,
  "cam_amm_team_rating": null,
  "cam_amm_team_count": 0,
  "cam_amm_team_conf": 0,
  "cam_amm_team_prov": 0,
  "cam_amm_team_pset": 0,
  "cam_herokills": 0,
  "cam_herodmg": 0,
  "cam_heroexp": 0,
  "cam_herokillsgold": 0,
  "cam_heroassists": 0,
  "cam_deaths": 0,
  "cam_goldlost2death": 0,
  "cam_secs_dead": 0,
  "cam_teamcreepkills": 0,
  "cam_teamcreepdmg": 0,
  "cam_teamcreepexp": 0,
  "cam_teamcreepgold": 0,
  "cam_neutralcreepkills": 0,
  "cam_neutralcreepdmg": 0,
  "cam_neutralcreepexp": 0,
  "cam_neutralcreepgold": 0,
  "cam_bdmg": 0,
  "cam_bdmgexp": 0,
  "cam_razed": 0,
  "cam_bgold": 0,
  "cam_denies": 0,
  "cam_exp_denied": 0,
  "cam_gold": 0,
  "cam_gold_spent": 0,
  "cam_exp": 0,
  "cam_actions": 0,
  "cam_secs": 0,
  "cam_consumables": 0,
  "cam_wards": 0,
  "cam_em_played": 0,
  "cam_level": 1,
  "cam_level_exp": 0,
  "cam_min_exp": 0,
  "cam_max_exp": 0,
  "cam_time_earning_exp": 0,
  "cam_bloodlust": 0,
  "cam_doublekill": 0,
  "cam_triplekill": 0,
  "cam_quadkill": 0,
  "cam_annihilation": 0,
  "cam_ks3": 0,
  "cam_ks4": 0,
  "cam_ks5": 0,
  "cam_ks6": 0,
  "cam_ks7": 0,
  "cam_ks8": 0,
  "cam_ks9": 0,
  "cam_ks10": 0,
  "cam_ks15": 0,
  "cam_smackdown": 0,
  "cam_humiliation": 0,
  "cam_nemesis": 0,
  "cam_retribution": 0,
  "acc_pub_skill": 1500,
  "acc_wins": 0,
  "acc_losses": 0,
  "acc_concedes": 0,
  "acc_concedevotes": 0,
  "acc_buybacks": 0,
  "acc_kicked": 0,
  "acc_amm_solo_rating": null,
  "acc_amm_solo_count": 0,
  "acc_amm_solo_conf": 0,
  "acc_amm_solo_prov": 0,
  "acc_amm_solo_pset": 0,
  "acc_amm_team_rating": null,
  "acc_amm_team_count": 0,
  "acc_amm_team_conf": 0,
  "acc_amm_team_prov": 0,
  "acc_amm_team_pset": 0,
  "acc_herokills": 0,
  "acc_herodmg": 0,
  "acc_heroexp": 0,
  "acc_herokillsgold": 0,
  "acc_heroassists": 0,
  "acc_deaths": 0,
  "acc_goldlost2death": 0,
  "acc_secs_dead": 0,
  "acc_teamcreepkills": 0,
  "acc_teamcreepdmg": 0,
  "acc_teamcreepexp": 0,
  "acc_teamcreepgold": 0,
  "acc_neutralcreepkills": 0,
  "acc_neutralcreepdmg": 0,
  "acc_neutralcreepexp": 0,
  "acc_neutralcreepgold": 0,
  "acc_bdmg": 0,
  "acc_bdmgexp": 0,
  "acc_razed": 0,
  "acc_bgold": 0,
  "acc_denies": 0,
  "acc_exp_denied": 0,
  "acc_gold": 0,
  "acc_gold_spent": 0,
  "acc_exp": 0,
  "acc_actions": 0,
  "acc_consumables": 0,
  "acc_wards": 0,
  "acc_em_played": 0,
  "acc_level": 1,
  "acc_level_exp": 0,
  "acc_min_exp": 0,
  "acc_max_exp": 0,
  "acc_time_earning_exp": 0,
  "acc_bloodlust": 0,
  "acc_doublekill": 0,
  "acc_triplekill": 0,
  "acc_quadkill": 0,
  "acc_annihilation": 0,
  "acc_ks3": 0,
  "acc_ks4": 0,
  "acc_ks5": 0,
  "acc_ks6": 0,
  "acc_ks7": 0,
  "acc_ks8": 0,
  "acc_ks9": 0,
  "acc_ks10": 0,
  "acc_ks15": 0,
  "acc_smackdown": 0,
  "acc_humiliation": 0,
  "acc_nemesis": 0,
  "acc_retribution": 0,
  "cam_cs_games_played": 0,
  "cam_cs_wins": 0,
  "cam_cs_losses": 0,
  "cam_cs_concedes": 0,
  "cam_cs_concedevotes": 0,
  "cam_cs_buybacks": 0,
  "cam_cs_discos": 0,
  "cam_cs_kicked": 0,
  "cam_cs_amm_solo_rating": null,
  "cam_cs_amm_solo_count": 0,
  "cam_cs_amm_solo_conf": 0,
  "cam_cs_amm_solo_prov": 0,
  "cam_cs_amm_solo_pset": 0,
  "cam_cs_amm_team_rating": null,
  "cam_cs_amm_team_count": 0,
  "cam_cs_amm_team_conf": 0,
  "cam_cs_amm_team_prov": 0,
  "cam_cs_amm_team_pset": 0,
  "cam_cs_herokills": 0,
  "cam_cs_herodmg": 0,
  "cam_cs_heroexp": 0,
  "cam_cs_herokillsgold": 0,
  "cam_cs_heroassists": 0,
  "cam_cs_deaths": 0,
  "cam_cs_goldlost2death": 0,
  "cam_cs_secam_cs_dead": 0,
  "cam_cs_teamcreepkills": 0,
  "cam_cs_teamcreepdmg": 0,
  "cam_cs_teamcreepexp": 0,
  "cam_cs_teamcreepgold": 0,
  "cam_cs_neutralcreepkills": 0,
  "cam_cs_neutralcreepdmg": 0,
  "cam_cs_neutralcreepexp": 0,
  "cam_cs_neutralcreepgold": 0,
  "cam_cs_bdmg": 0,
  "cam_cs_bdmgexp": 0,
  "cam_cs_razed": 0,
  "cam_cs_bgold": 0,
  "cam_cs_denies": 0,
  "cam_cs_exp_denied": 0,
  "cam_cs_gold": 0,
  "cam_cs_gold_spent": 0,
  "cam_cs_exp": 0,
  "cam_cs_actions": 0,
  "cam_cs_secs": 0,
  "cam_cs_consumables": 0,
  "cam_cs_wards": 0,
  "cam_cs_em_played": 0,
  "cam_cs_level": 1,
  "cam_cs_level_exp": 0,
  "cam_cs_min_exp": 0,
  "cam_cs_max_exp": 0,
  "cam_cs_time_earning_exp": 0,
  "cam_cs_bloodlust": 0,
  "cam_cs_doublekill": 0,
  "cam_cs_triplekill": 0,
  "cam_cs_quadkill": 0,
  "cam_cs_annihilation": 0,
  "cam_cs_ks3": 0,
  "cam_cs_ks4": 0,
  "cam_cs_ks5": 0,
  "cam_cs_ks6": 0,
  "cam_cs_ks7": 0,
  "cam_cs_ks8": 0,
  "cam_cs_ks9": 0,
  "cam_cs_ks10": 0,
  "cam_cs_ks15": 0,
  "cam_cs_smackdown": 0,
  "cam_cs_humiliation": 0,
  "cam_cs_nemesis": 0,
  "cam_cs_retribution": 0,
  "name": null,
  "rank": "None",
  "favHero1": null,
  "favHero1Time": 0,
  "favHero1_2": null,
  "favHero2": null,
  "favHero2Time": 0,
  "favHero2_2": null,
  "favHero3": null,
  "favHero3Time": 0,
  "favHero3_2": null,
  "favHero4": null,
  "favHero4Time": 0,
  "favHero4_2": null,
  "favHero5": null,
  "favHero5Time": 0,
  "favHero5_2": null,
  "dice_tokens": 1,
  "season_level": 0,
  "slot_id": 5,
  "my_upgrades": {
    "0": "m.allmodes.pass",
    "1": "h.AllHeroes.Hero",
    "2": "m.Super-Taunt"
  },
  "selected_upgrades": {
    "0": "ai.DW Hardcore Marathoners",
    "1": "av.8-Bit Announcer Pack",
    "2": "c.8bit_courier",
    "3": "cc.pixelpower",
    "4": "cs.Doge Anubis",
    "5": "sc.punk_circle_upgrade",
    "6": "t.Gamer_Rage_Taunt",
    "7": "te.8 Bit TP Effect",
    "8": "w.independence_day_ward"
  },
  "game_tokens": 0,
  "my_upgrades_info": {},
  "creep_level": 0,
  "timestamp": 1668512501,
  "matchIds": null,
  "matchDates": null,
  "k_d_a": "0/0/0",
  "avgGameLength": 0,
  "avgXP_min": null,
  "avgDenies": 0,
  "avgCreepKills": 0,
  "avgNeutralKills": 0,
  "avgActions_min": 0,
  "avgWardsUsed": 0,
  "quest_stats": {
    "error": {
      "quest_status": 0,
      "leaderboard_status": 0
    }
  },
  "highest_level_current": 0,
  "current_level": 0,
  "level_percent": 0,
  "season_id": 22,
  "prev_seasons_cam_games_played": 0,
  "prev_seasons_cam_discos": 0,
  "latest_season_cam_games_played": 0,
  "latest_season_cam_discos": 0,
  "curr_season_cam_games_played": 0,
  "curr_season_cam_discos": 0,
  "prev_seasons_cam_cs_games_played": 0,
  "prev_seasons_cam_cs_discos": 0,
  "latest_season_cam_cs_games_played": 0,
  "latest_season_cam_cs_discos": 0,
  "curr_season_cam_cs_games_played": 0,
  "curr_season_cam_cs_discos": 0,
  "con_reward": {
    "old_lvl": 0,
    "curr_lvl": 0,
    "next_lvl": 0,
    "require_rank": 0,
    "need_more_play": 0,
    "percentage_before": "0.00",
    "percentage": "0.00"
  },
  "vested_threshold": 5,
  "0": true
}